### DISCUSSION TITLE
Error: pages with `getServerSideProps` can not be exported

### DISCUSSION THREAD

**Discussion by pawel2105** [vercel/next.js#11945] [Help] [answered]
I am trying to get a page to use `getServerSideProps` but when I try to run a build on **Netlify** then I get the following error:

```
Error occurred prerendering page "/profile/[username]". 
Read more: https://err.sh/next.js/prerender-error:
Error: Error for page /profile/[username]: pages with `getServerSideProps` can not be exported.
See more info here: https://err.sh/next.js/gssp-export
```

I'm not sure why I'm getting this error since it is a page, and google returns no results for me for an exact search for `pages with 'getServerSideProps' can not be exported`. 

Here is the content for `/pages/profile/[username].js`:

```
import React, { Component } from "react";
import { connect } from "react-redux";
import { withRouter } from "next/router";
import * as profilesModule from "../../redux/modules/profiles";
import { fetchProfile } from "../../redux/apiModules/profiles";

class ProfilePage extends Component {
  componentDidMount() {
    const { query } = this.props.router;
    const { username } = query;

    this.props.fetchProfile({ username });
  }

  render() {
    const { profile } = this.props;

    return <div>Hello {profile.username}</div>;
  }
}

const mapStateToProps = (state) => ({ profile: state.profiles.profileData });
const mapDispatchToProps = (dispatch) => ({
  fetchProfile: (username) => dispatch(profilesModule.fetchProfile(username)),
});

export async function getServerSideProps({ params }) {
  // I can see 2 request to the EXTERNAL API for this username on each page load
  // Assuming this is correct (1 for getServerSideProps and 1 for componentDidMount)
  const profile = await fetchProfile({ username: params.username });
  return { props: { profile: profile.data } };
}

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(ProfilePage)
);
```

**Could the problem be the redux wrapping?**

Nope. I commented out the redux stuff and the build error persists:

```
import React, { Component } from "react";
import { withRouter } from "next/router";
import { fetchProfile } from "../../redux/apiModules/profiles";

class ProfilePage extends Component {
  render() {
    // this.props.profile is undefined in the browser's console
    console.log(this.props.profile);
    
    const { profile = {} } = this.props;
    return <div>Hello {profile.username}</div>;
  }
}

export async function getServerSideProps({ params }) {
  // I can see 1 request to the EXTERNAL API for this username on each page load
  const profile = await fetchProfile({ username: params.username });
  return { props: { profile: profile.data } };
}

export default withRouter(ProfilePage);
```

**Could the problem be the withRouter wrapping?**

I don't know. I'm not sure how to extract the query param without wrapping the page with withRouter.

**Additional details that might be relevant:**

I'm using `"next": "^9.3.5"`.

My Netlify build command is `npm run build && npm run export && npm run postexport`. Postexport is just to copy over a _redirects file for Netlify.

Since Netlify is hosting this app, if I understand correctly only `next export` works so `next start` is not an option.

My `App` component is customized and wrapped with `withRedux` but does not make use of `getInitialProps`:

```
import App from "next/app";
import Router from "next/router";
import React from "react";
import { Provider } from "react-redux";
import NProgress from "nprogress";
import withRedux from "next-redux-wrapper";
import withReduxSaga from "next-redux-saga";

// custom CSS
import GlobalStyles from "../components/globals";

// redux store
import createStore from "../redux/store";

// progress bar
Router.events.on("routeChangeStart", () => NProgress.start());
Router.events.on("routeChangeComplete", () => NProgress.done());
Router.events.on("routeChangeError", () => NProgress.done());

class MyApp extends App {
  render() {
    const { Component, pageProps, store } = this.props;
    return (
      <Provider store={store}>
        <Component {...pageProps} />
        <GlobalStyles />
      </Provider>
    );
  }
}

export default withRedux(createStore)(withReduxSaga(MyApp));
```

My `next.config.js` file looks like:

```
const withImages = require("next-images");

const determineBaseUrl = () => {
  // stuff for API base URL determination
}

const baseUrl = determineBaseUrl();

module.exports = withImages({ env: { baseUrl } });

```

Am I missing something super obvious?

**Comment by sergiocardoso** [comment]
same here =(

**Comment by ijjk** [comment, marked as answer]
Hi, if you visit [the link](https://err.sh/next.js/gssp-export) provided in the error it provides additional information for why you are seeing it. To recap what that link mentions, we error when trying to run `next export` (which is required to deploy on Netlify) with a page with `getServerSideProps` because it defeats the purpose of the method since it can't be called on every request if you statically export the page. 

Instead you should be leveraging `getStaticProps` to load data at build time so that it's available when statically exporting a page. If you instead are trying to deploy your app with `getServerSideProps` and be able to fetch data on every request as the method is intended to work you can try deploying on a platform that supports it out of the box like [ZEIT](https://zeit.co)

**Reply by byte-artisan** [reply]
I got this same problem with ZEIT. It breaks on build both on ZEIT and on local machine.
The code can be view [here ](https://vercel.com/byte-artisan/next-test/exqd2uw5u)at ZEIT, it is public.

**Reply by timneutkens** [reply]
```
Error occurred prerendering page "/404". Read more: https://err.sh/next.js/prerender-error:
22:50:09.186  
Error: Error for page /_error: pages with `getServerSideProps` can not be exported. See more info here: https://err.sh/next.js/gssp-export
```

You have a `pages/_error.js` which has:

```
import React from 'react';



export async function getServerSideProps(context) {
  return {
    props: {value:'x-error'}, // will be passed to the page component as props
  }
}


function Error({ statusCode }) {
  return (
    <p>
      {statusCode ? `An error ${statusCode} occurred on server` : 'An error occurred on client'}
    </p>
  );
}



export default Error;

```

pages/_error is only rendered in case there is an error so it can't use `getServerSideProps` as it has to be rendered immediately when an error happens.

**Reply by byteartisan** [reply]
The possibility of doing 301 redirects if choose to could pay the extra time to load the page, in my opinion.

**Reply by tiavina-mika** [reply]
"Instead you should be leveraging getStaticProps to load data at build time so that it's available when statically exporting a page" but how to do that?

**Comment by pawel2105** [comment]
Whoops, sorry about the unmarking, I thought I _was_ marking as correct. Also, tx for the explanation.

**Comment by sarink** [comment]
I am getting this error on the `_error` page. I am not using `next export` and I am not using `getStaticProps`. Why?

**Comment by tedescodev** [comment]
Same problems for us. A big mistake to have used Next.

**Comment by avin1208** [comment]
@tedescodev use getinitialProps instead of staticProps

and  return {
        props: {
            anyName : yourState ? yourState : "No Data Available",
        }
    }
    
    pass as a stirng in condtional statement . then it will work in deployment and production
